### ISSUE TITLE
PKCS12 import fails with "MAC verification failed" error using correct password

### ISSUE THREAD

**Issue by ackintosh** [kornelski/rust-security-framework#216]
I consistently encounter the error `MAC verification failed during PKCS12 import (wrong password?)`, even though the password is correct.

```rust
    // The following commands are used to generate the key and certificate.
    // openssl genrsa -out server.key 2048
    // openssl req -new -key server.key -out server.csr
    // openssl x509 -req -days 365 -in server.csr -signkey server.key -out server.crt
    // openssl pkcs12 -export -out server.p12 -inkey server.key -in server.crt -password pass:bark
    #[test]
    fn test() {
        let path = PathBuf::from(env::var("CARGO_MANIFEST_DIR").unwrap()).join("server.p12");
        let mut buf = Vec::new();
        File::open(path)
            .unwrap()
            .read_to_end(&mut buf)
            .unwrap();
        
        let mut import_opts = Pkcs12ImportOptions::new();
        let imports = import_opts.passphrase("bark").import(&buf);
        match imports {
            Ok(_) => {
                println!("Import successful");
            }
            Err(error) => {
                println!("{:?}", error);
            }
        }
    }
```

The error encountered is:

```
Error { code: -25264, message: "MAC verification failed during PKCS12 import (wrong password?)" }
```

To verify, I confirmed that I can successfully extract the certificate and key from server.p12 using the passphrase with the following OpenSSL command:

```bash
openssl pkcs12 -in server.p12 -clcerts -nokeys -out public-key.crt
openssl pkcs12 -in server.p12 -nocerts -nodes -out private-key.pem
```

macOS  Sonoma 14.6.1
security-framework 3.0.0

**Comment by kornelski** [comment]
Do you have a working equivalent in C or ObjC?

**Comment by ackintosh** [comment]
I'm new to C (and ObjC), so I generated the following C code with the help of ChatGPT.

```c
#include <stdio.h>
#include <openssl/pkcs12.h>
#include <openssl/err.h>
#include <openssl/ssl.h>

int main() {
    const char *filename = "server.p12";
    const char *password = "bark";
    FILE *fp = fopen(filename, "rb");
    if (!fp) {
        perror("Cannot open file");
        return 1;
    }

    PKCS12 *p12 = d2i_PKCS12_fp(fp, NULL);
    fclose(fp);
    if (!p12) {
        fprintf(stderr, "Failed to read PKCS12 file\n");
        ERR_print_errors_fp(stderr);
        return 1;
    }

    EVP_PKEY *pkey = NULL;
    X509 *cert = NULL;
    STACK_OF(X509) *ca = NULL;

    OpenSSL_add_all_algorithms();
    ERR_load_crypto_strings();

    if (!PKCS12_parse(p12, password, &pkey, &cert, &ca)) {
        fprintf(stderr, "Failed to parse PKCS12 file\n");
        ERR_print_errors_fp(stderr);
        PKCS12_free(p12);
        return 1;
    }

    printf("Import successful\n");

    // Free memory
    PKCS12_free(p12);
    EVP_PKEY_free(pkey);
    X509_free(cert);
    if (ca) {
        sk_X509_pop_free(ca, X509_free);
    }

    return 0;
}
```

I ran this, and it returned `Import successful`.

**Comment by kornelski** [comment]
That's using OpenSSL, not Apple Security framework. 

I'm asking for Apple code, because it's not clear whether the problem is with the Rust wrapper, or whether Apple just doesn't support it properly.

**Comment by ackintosh** [comment]
I re-generated the following C code to use Apple Security framework, and the error  was reproduced. So It looks Apple just doesn't support it properly.

<details><summary> test_p12.c </summary>
<p>

```c
#include <stdio.h>
#include <stdlib.h>
#include <CoreFoundation/CoreFoundation.h>
#include <Security/Security.h>

int main() {
    const char *filename = "server.p12";
    const char *password = "bark";

    // Read the PKCS#12 file
    CFDataRef pkcs12Data = NULL;
    {
        FILE *fp = fopen(filename, "rb");
        if (!fp) {
            perror("Cannot open file");
            return 1;
        }
        fseek(fp, 0, SEEK_END);
        long fileSize = ftell(fp);
        rewind(fp);

        unsigned char *buffer = (unsigned char *)malloc(fileSize);
        if (fread(buffer, 1, fileSize, fp) != fileSize) {
            perror("Failed to read file");
            fclose(fp);
            free(buffer);
            return 1;
        }
        fclose(fp);

        pkcs12Data = CFDataCreate(NULL, buffer, fileSize);
        free(buffer);
    }

    // Set PKCS#12 import options
    CFStringRef passwordStr = CFStringCreateWithCString(NULL, password, kCFStringEncodingUTF8);
    const void *keys[] = { kSecImportExportPassphrase };
    const void *values[] = { passwordStr };
    CFDictionaryRef options = CFDictionaryCreate(NULL, keys, values, 1, NULL, NULL);

    // Import the PKCS#12 file
    CFArrayRef items = NULL;
    OSStatus status = SecPKCS12Import(pkcs12Data, options, &items);

    // Process the result
    if (status == errSecSuccess && CFArrayGetCount(items) > 0) {
        CFDictionaryRef identityAndTrust = CFArrayGetValueAtIndex(items, 0);
        SecIdentityRef identity = (SecIdentityRef)CFDictionaryGetValue(identityAndTrust, kSecImportItemIdentity);
        SecCertificateRef certificate = NULL;
        SecKeyRef privateKey = NULL;

        // Retrieve the certificate
        if (identity) {
            OSStatus certStatus = SecIdentityCopyCertificate(identity, &certificate);
            if (certStatus == errSecSuccess) {
                CFStringRef certSummary = SecCertificateCopySubjectSummary(certificate);
                char certName[256];
                CFStringGetCString(certSummary, certName, sizeof(certName), kCFStringEncodingUTF8);
                printf("Certificate Subject: %s\n", certName);
                CFRelease(certSummary);
                CFRelease(certificate);
            }
        }

        // Retrieve the private key
        if (identity) {
            OSStatus keyStatus = SecIdentityCopyPrivateKey(identity, &privateKey);
            if (keyStatus == errSecSuccess) {
                printf("Private key imported successfully\n");
                CFRelease(privateKey);
            }
        }

        printf("Import successful\n");
    } else {
        CFStringRef errorMessage = SecCopyErrorMessageString(status, NULL);
        char errorStr[256];
        CFStringGetCString(errorMessage, errorStr, sizeof(errorStr), kCFStringEncodingUTF8);
        fprintf(stderr, "Failed to import PKCS12 file: %s (Error code: %d)\n", errorStr, (int)status);
        CFRelease(errorMessage);
    }

    // Release memory
    if (items) CFRelease(items);
    CFRelease(options);
    CFRelease(passwordStr);
    CFRelease(pkcs12Data);

    return 0;
}
```

</p>
</details> 

```bash
$ gcc -o test_p12 test_p12.c -framework Security -framework CoreFoundation
$ ./test_p12
Failed to import PKCS12 file: MAC verification failed during PKCS12 import (wrong password?) (Error code: -25264)
```

**Comment by ackintosh** [comment]
I found this post from Apple Developer Forums.

[SecPKCS12Import is failing to import P12 certificate. | Apple Developer Forums](https://forums.developer.apple.com/forums/thread/697030)
 
OpenSSL has changed the default algorithms for pkcs12 since v3 ([CHANGELOG](https://github.com/openssl/openssl/commit/762970bd686c4aa8ea7169e7f76d5a4ce665da93#diff-d975bf659606195d2165918f93e1cf680ef68ea3c9cab994f033705fea8238b2R129-R139)). `SecPKCS12Import` is not compatible with the default `openssl pkcs12` encoding algorithm in OpenSSL version 3. Therefore, p12 files need to be created using `openssl pkcs12 -legacy` to import successfully.

I tried the C code with a p12 file created using `-legacy` option, and it worked successfully. 

```bash
$ openssl pkcs12  -legacy -export -out server.p12 -inkey server.key -in server.crt -password pass:bark
$ ./test_p12 
Certificate Subject: Example Corp
Private key imported successfully
Import successful
```

The Rust code also ran successfully. 🎉
